vmm/src/lib.rs

```diff
@@ -605,58 +574,73 @@ impl EpollContext {
             epoll::ControlOptions::EPOLL_CTL_DEL,
             libc::STDIN_FILENO,
             epoll::Event::new(epoll::Events::EPOLLIN, self.stdin_index),
-        )
-        .map_err(Error::EpollFd);
+        );
         self.dispatch_table[self.stdin_index as usize] = None;
 
         Ok(())
     }
 
-    fn add_event<T>(&mut self, fd: T, token: EpollDispatch) -> Result<EpollEvent<T>>
-    where
-        T: AsRawFd,
-    {
+    /// Given a file descriptor `fd`, and an EpollDispatch token `token`,
+    /// associate `token` with an `EPOLLIN` event for `fd`, through the
+    /// `dispatch_table`.
+    fn add_event<T: AsRawFd>(&mut self, fd: &T, token: EpollDispatch) -> Result<()> {
+        // The index in the dispatch where the new token will be added.
         let dispatch_index = self.dispatch_table.len() as u64;
+
+        // Add a new epoll event on `fd`, associated with index
+        // `dispatch_index`.
         epoll::ctl(
             self.epoll_raw_fd,
             epoll::ControlOptions::EPOLL_CTL_ADD,
             fd.as_raw_fd(),
             epoll::Event::new(epoll::Events::EPOLLIN, dispatch_index),
         )
         .map_err(Error::EpollFd)?;
+
+        // Add the associated token at index `dispatch_index`
         self.dispatch_table.push(Some(token));
 
-        Ok(EpollEvent { fd })
+        Ok(())
     }
 
-    fn allocate_tokens(&mut self, count: usize) -> (u64, Sender<Box<EpollHandler>>) {
+    /// Allocates `count` dispatch tokens, simultaneously registering them in
+    /// `dispatch_table`. The tokens will be associated with a device.
+    /// This device's handler will be added to the end of `device_handlers`.
+    /// This returns the index of the first token, and a channel on which to
+    /// send an epoll handler for the relevant device.
+    fn allocate_tokens_for_device(&mut self, count: usize) -> (u64, Sender<Box<EpollHandler>>) {
         let dispatch_base = self.dispatch_table.len() as u64;
         let device_idx = self.device_handlers.len();
         let (sender, receiver) = channel();
 
-        for x in 0..count {
-            self.dispatch_table.push(Some(EpollDispatch::DeviceHandler(
+        self.dispatch_table.extend((0..count).map(|index| {
+            Some(EpollDispatch::DeviceHandler(
                 device_idx,
-                x as DeviceEventT,
-            )));
-        }
-
+                index as DeviceEventT,
+            ))
+        }));
         self.device_handlers.push(MaybeHandler::new(receiver));
 
         (dispatch_base, sender)
     }
 
-    fn allocate_virtio_tokens<T: EpollConfigConstructor>(
+    /// Allocate tokens for a virtio device, as with `allocate_tokens_for_device`,
```

Comment:
This is a weird function as it has side-effects. We can add a new issue with it and use the label Technical Dept. A function called `allocate_tokens...` should allocate tokens and not have other side effects.

Reply:
Done in issue #1225